Add `texttemplate` attribute to `shape.label`

emilykl opened the pull request:
Resolves #6511 

Add `texttemplate` attribute to `shape.label` and `newshape.label`.

`texttemplate` supports insertion of shape variables using `"%{variable}"` syntax, with d3 number and date formatting. A single multiplication or division operation is also supported, to allow for unit conversions. Shape label displays correct values for variables used in `texttemplate`, and update values live as shape is moved or resized.

Supported variables:
- Raw variables (from shape definition):
  - `x0`
  - `x1`
  - `y0`
  - `y1`
Calculated variables:
  - `xcenter` (calculated as `(x0+x1)/2`)
  - `ycenter` (calculated as `(y0+y1)/2`)
  - `dx` (calculated as `x1-x0`)
  - `dy` (calculated as `y1-y0`)
  - `width` (calculated as `abs(x1-x0)`)
  - `height` (calculated as `abs(y1-y0)`)
  - `length` (calculated as `sqrt(dx^2+dy^2)`) -- for lines only
  - `slope` (calculated as `(y1-y0)/(x1-x0)`)

Behavior for `log`, `date`, and `category` axes:
- Raw variables use raw data values from shape definition (datetimes for `date` axes and strings for `category` axes)
- Calculated variables use linearized values (log values for `log` axes, milliseconds for `date` axes, integers for `category` axes) 
  - `xcenter` and `ycenter` are converted back to data values after calculation

This PR does not cover rendering of the shape text during mousedown of the initial draw -- due to additional complexity, that will be covered in a separate issue, captured by #6540.

## API

- `shape.label.texttemplate`: Template string used for rendering the new shape's label. Note that this will override `text`. Variables are inserted using `"%{variable}"`, for example `"x0: %{x0}"`. Numbers are formatted using d3-format's syntax `"%{variable:d3-format}"`, for example `"Price: %{x0:$.2f}"`. See https://github.com/d3/d3-format/tree/v1.4.5#d3-format for details on the formatting syntax. 

  Dates are formatted using d3-time-format's syntax `"%{variable|d3-time-format}"`, for example `"Day: %{x0|%A}"`. See: https://github.com/d3/d3-time-format/tree/v2.2.3#locale_format 

  A single multiplication or division operation may be applied to numeric variables, and combined with d3 number formatting, for example `"Length in cm: %{x0*2.54}"`, `"%{slope*60:.1f} meters per second."` 

  For log axes, variable values are given in log units. For date axes, x/y coordinate variables and center variables use datetimes, while all other variable values use values in ms. 

  Finally, the template string has access to variables `x0`, `x1`, `y0`, `y1`, `slope`, `dx`, `dy`, `width`, `height`, `length`, `xcenter` and `ycenter`.

## Partnership

Development of this feature is sponsored by Volkswagen's Center of Excellence for Battery Systems.

<p float="left" align="center" >
<img src="https://user-images.githubusercontent.com/42301846/207075484-2c77d1e1-f0b5-43da-89bf-66ac73c50007.png" width=40% aspect-ratio=3/10 >
</p>

emilykl commented:
Currently working for "easy" cases. Edge cases I haven't tackled yet:

- Paths (no x0/x1/y0/y1)
- Shapes where one of the coordinates is a datetime string
- Divide by zero errors
- "slope" for circles and rectangles -- should we allow it?

emilykl commented:
Here is a codepen demonstrating the current behavior: https://codepen.io/emily_plotly/pen/eYLrKNV

emilykl commented:
Thanks @alexcjohnson !

> * `dx===x1-x0`
> * `dy===y1-y0`
> * `length===sqrt(dx^2+dy^2)`
> * `xcenter===(x0+x1)/2` or maybe just `x`
> * `ycenter===(y0+y1)/2` or `y`

These should all be pretty easy. For the first two, would `width` and `height` be better names than `dx` and `dy`?

> Let's not support paths for now.

Fine with me. How should we handle the case where a user sets a `texttemplate` for a path? Not display the text at all? Display the template string unaltered? Display `undefined` for the variables?

alexcjohnson commented:
> For the first two, would `width` and `height` be better names than `dx` and `dy`?

yes 😁 

> How should we handle the case where a user sets a `texttemplate` for a path?

Do whatever we do today if you make a `texttemplate` or `hovertemplate` in another context that has invalid fields.

archmoj commented:
> Nice! If it's easy there are a few more that occur to me might be useful. We can leave them for later too though:
> 
> * `dx===x1-x0`
> * `dy===y1-y0`
> * `length===sqrt(dx^2+dy^2)`
> * `xcenter===(x0+x1)/2` or maybe just `x`
> * `ycenter===(y0+y1)/2` or `y`
> 

Instead of `length` why not using `centroid` or `cen`?
That would also have a meaning for paths.

On another note, perhaps `atan2((x1-x0)/(y1-y0))` converted to degrees would also be very useful. 
It doesn't have problem with Infinity as well as big and small numbers.

Also concerning `dx` and `dy` shouldn't the absolute be printed?
Or perhaps we could add `xlen` and `ylen`.

alexcjohnson commented:
Ah interesting - `width` and `height` would be only positive but `dx` and `dy` would be signed. So perhaps there’s room for all of those.

`angle` is interesting, sure. I’ll note that both `angle` and `length` are meaningful only on a space with a well-defined metric, ie x and y have matching units, and they’ll look funny if you don’t enforce `scaleratio = 1`. There would also be cases where they might *seem* correct but they aren’t, like if you plot latitude and longitude on cartesian axes rather than a map. Not a reason to not do them, just a note of caution we might include in our docs. 

Let’s stay away from paths please. It’s going to be a rabbit hole dealing with straight lines, arcs, and two orders of Bézier curves, plus open and closed paths, and nobody has asked for it. If and when we do that we can add `centroid` or whatever, but no need to tailor the fields that apply to simple shapes to try and match paths, it’ll just make the normal cases more confusing.

emilykl commented:
Sure, I can see an argument for having both `dy`/`dx` (signed) and `width`/`height` (unsigned).

Is there ever an instance where the x and y axes are swapped, i.e. y is horizontal and x is vertical? That would be weird but just double-checking whether `width` and `height` variable names make sense in all cases.

I can't think of a use-case for `angle` that isn't affected by the issues @alexcjohnson mentioned so I'd argue against it for now.

JulianWgs commented:
Very cool discussion going on!

@alexcjohnson 
> Seems like we should handle absolute and relative differently here: for {x|y}{0|1} we should use the date directly, but support d3 datetime format specifiers. For slope (and, later, if we add dx, dy, or length) it should use milliseconds.

Why would you opt for milliseconds? I know that in many programming language including JavaScript milliseconds is the default timedelta unit. However in Python it is the much human friendly second 😄 Would it add too much complexity to have different units for plotly.js and plotly.py or just use seconds all together? Also since it is only for displaying we dont encounter floating point arithmetic issues.

@emilykl Could you please update the pull request description, whenever we finalized the set of supported variables. Also it might be worthwhile to split the pull request into the common and not so common cases, but I dont have a strong opinion about this 😄 This may include log axis support, since although the use case you describe seams really elegant I dont know how often this is actually is needed.

I would prefer `xcenter`, `ycenter`, because many programs dont anchor their shapes in the center, so just `x` and `y` might be confusing.

emilykl commented:
@JulianWgs Thank you, great to have your feedback.

For the final list of variables, @archmoj are you okay with the following? If there's any dispute about any of these I'd argue for leaving them out of this PR rather than prolonging the discussion, since it's trivial to add more variables later.
- `x0`
- `x1`
- `y0`
- `y1`
- `xcenter` (calculated as `(x0+x1)/2`)
- `ycenter` (calculated as `(y0+y1)/2`)
- `dx` (calculated as `x1-x0`)
- `dy` (calculated as `y1-y0`)
- `width` (calculated as `abs(x1-x0)`)
- `height` (calculated as `abs(y1-y0)`)
- `length` (calculated as `sqrt(dx^2+dy^2)`)
- `slope` (calculated as `(y1-y0)/(x1-x0)`)

Regarding time units, I believe Plotly.js uses milliseconds internally for datetime representations, and using a different unit only for shape labels has the potential to create some confusion -- I will verify internally though whether there's precedent for this question. 

@archmoj @alexcjohnson Would it be reasonable to exclude slope calculations for log axes for the moment? Unless you think the correct slope is obvious in all cases and it's a very simple implementation.

archmoj commented:
> 

Let's leave `log` axes for now.
IMHO `width` & `height` could confusing as they may point to pixel coordinate.
I'd also avoid `length` as it could be mixed with the `length` of arrays in JavaScript. So could we use `len` instead?
Also for `center` we could use `cen` or `mid`.
So here is my suggestion:
 
- `x0`
- `x1`
- `y0`
- `y1`
- `xmid` (calculated as `(x0+x1)/2`)
- `ymid` (calculated as `(y0+y1)/2`)
- `dx` (calculated as `x1-x0`)
- `dy` (calculated as `y1-y0`)
- `xlen` (calculated as `abs(x1-x0)`)
- `ylen` (calculated as `abs(y1-y0)`)
- `len` (calculated as `sqrt(dx^2+dy^2)`)
- `slope` (calculated as `(y1-y0)/(x1-x0)`)

Finally instead of `slope` could we use `tan`? 
If so, then we could similarly add `atan` for the angle of inclination.

alexcjohnson commented:
Let’s keep `slope`, and if we add angle let’s just call it `angle`, even though if you stretch the axes (or invert them!) the angle on-screen will be different (the same can be said of slope, though typically a slope is given with units so this is clearer). 

Personally I like `width`, `height`, and `length`, I think if I were not reading the docs but just guessing those are among the first I would guess whereas `len` would take a while and I might never guess `xlen` and `ylen`. But I feel super strongly about that, with docs all of those names work.

To me, the behavior on log axes is clear (one unit of difference on a log scale is one power of ten in the data, again the same as our internal representation), but since nobody is requesting it right now we can leave it for later. 

Re: milliseconds - we have a bunch of places in plotlyjs where time differences are already specified in milliseconds, so I don’t think it’s a good idea to change just this one. At some point we could try to add a units field somewhere, then people could specify seconds, days, or whatever. But absent that I think we need to leave it as milliseconds.

JulianWgs commented:
I would always prefer English words over mathematical or software expressions (but for example dx and dy are fine), but this might be due to my Python background.

`len(array)` in Python gives you the length of an array, so there will always be confusion depending on background, so I prefer `length`. Although [distance](https://en.wikipedia.org/wiki/Euclidean_distance) might be a good fit as well.

Slope is a must for our use case.

@emilykl Everything else on your list is fine for me as well. I haven't checked the code yet. Have you started the implementation? If we need further discussion on naming I would propose not to wait for the discussion to end, but start implementing as soon as possible :)

emilykl commented on src/plots/template_attributes.js:
```diff
@@ -33,9 +50,9 @@ function describeVariables(extra) {
         descPart = descPart + 'Finally, the template string has access to ';
         if(keys.length === 1) {
-            descPart = 'variable ' + quotedKeys[0];
+            descPart = descPart + 'variable ' + quotedKeys[0];
```
Note: I fixed this bug in the `describeVariables()` function, which is the reason for the large number of changes in `plot-schema.json`, since a number of `hovertemplate` and `texttemplate` docstrings were modified. In all cases I think the new string is an improvement.

archmoj replied:
Good catch. Thank you for the fix!

emilykl commented:
@archmoj @alexcjohnson This is ready for your review. All tests are passing except for `webgl-jasmine`, unrelated.

@JulianWgs I've updated the PR description to reflect the final list of supported variables.

Codepen here: https://codepen.io/emily_plotly/pen/rNZbLqx?editors=1010

===

My one remaining doubt is that it's not actually obvious to me that giving variable values in log units is the correct behavior for log axes. It has the odd consequence for `x0`/`x1`/`y0`/`y1` where the displayed variable value is different from the value seen on the axis -- see the below image for an example. The raw value for `x0` is `10`, and that's what's seen on the axis as well, but the value displayed in the text template is `1`, and likewise for the other variables.

We could in theory use raw values only for `x0`/`x1`/`y0`/`y1`, and log values for the rest, but I'm not sure if that's correct either, and has the disadvantage of being inconsistent.

If you feel confident that using log values always is correct, despite the weirdness, I'm OK with it. Alternatively, we could disable variable support for log axes for the moment.

![Screen Shot 2023-03-28 at 9 35 06 PM](https://user-images.githubusercontent.com/4672118/228403712-3a35e183-a757-47c5-bb90-ce69c5fd4b85.png)

archmoj commented on src/components/shapes/label_texttemplate.js:
```diff
@@ -0,0 +1,28 @@
+'use strict';
+
+module.exports = {
+    shapeLabelTexttemplateVars: {
```
Let's drop `shapeLabelTexttemplateVars` and simplify this export now that there is only one object being exported.
```js
module.exports = {
  x0: ...,
  x1: ...,
  ...
  
  Then the `require` statements would be
  ```js
  var shapeLabelTexttemplateVars = require('./label_texttemplate');
  ``` 
};

archmoj commented on src/components/shapes/label_texttemplate.js:
```diff
@@ -0,0 +1,28 @@
+
+module.exports = {
+    shapeLabelTexttemplateVars: {
+        x0: (function(shape, _xa) { return (_xa.type === 'date') ? shape.x0 : _xa.d2l(shape.x0); }),
```
We could rename `_xa` to `xa`, etc. in these functions here.

archmoj commented on src/lib/index.js:
```diff
@@ -1074,6 +1074,26 @@ lib.texttemplateString = function() {
+function multDivParser(inputStr) {
+    var match = inputStr.match(MULT_DIV_REGEX);
+    if(match) return { key: match[1], op: match[2], number: Number(match[3]) };
+    else return { key: inputStr, op: null, number: null };
```
```suggestion
    return { key: inputStr, op: null, number: null };
```

archmoj commented on src/lib/index.js:
```diff
@@ -1145,6 +1176,14 @@ function templateFormatString(string, labels, d3locale) {
         }
 
+        // Apply mult/div operation (if applicable)
+        if(parsedOp) {
```
Let's be more strict here e.g.:
```js
        if(parsedOp !== null) {
```
Or even better would be
```suggestion
        if(parsedOp === '*' || parsedOp === '/') {
```

archmoj commented on src/lib/index.js:
```diff
@@ -1145,6 +1176,14 @@ function templateFormatString(string, labels, d3locale) {
 
+        // Apply mult/div operation (if applicable)
+        if(parsedOp) {
+            value = (value === undefined) ? undefined : {
```
If you test against ` value !== undefined` in the upper condition, this logic would be simplified.

archmoj commented:
@emilykl very nice PR :medal_military: 
I added some minor comments.

archmoj commented on test/image/mocks/zz-text_on_shapes_texttemplate.json:
```diff
@@ -0,0 +1,334 @@
+            "drawopenpath",
+            "drawclosedpath",
+            "drawcircle",
+            "drawrect",
```
This is a beautiful mock!
BTW I noticed if one selects `drawrect` or `drawcircle` and then start drawing new shapes; it prints out some `slope` values which it shouldn't.
So we need to fix that.

emilykl replied:
yeah initially I was thinking we should just allow it, but the more I think about it, the more it seems like we should not allow `slope` for `rect` and `circle`, it just seems kind of weird.

archmoj commented:
> @archmoj @alexcjohnson This is ready for your review. All tests are passing except for `webgl-jasmine`, unrelated.
> 
> @JulianWgs I've updated the PR description to reflect the final list of supported variables.
> 
> Codepen here: https://codepen.io/emily_plotly/pen/rNZbLqx?editors=1010
> 
> ===
> 
> My one remaining doubt is that it's not actually obvious to me that giving variable values in log units is the correct behavior for log axes. It has the odd consequence for `x0`/`x1`/`y0`/`y1` where the displayed variable value is different from the value seen on the axis -- see the below image for an example. The raw value for `x0` is `10`, and that's what's seen on the axis as well, but the value displayed in the text template is `1`, and likewise for the other variables.
> 
> We could in theory use raw values only for `x0`/`x1`/`y0`/`y1`, and log values for the rest, but I'm not sure if that's correct either, and has the disadvantage of being inconsistent.
> 
> If you feel confident that using log values always is correct, despite the weirdness, I'm OK with it. Alternatively, we could disable variable support for log axes for the moment.
> 
> ![Screen Shot 2023-03-28 at 9 35 06 PM](https://user-images.githubusercontent.com/4672118/228403712-3a35e183-a757-47c5-bb90-ce69c5fd4b85.png)

@emilykl to fix/improve `log` case, you may consider picking https://github.com/plotly/plotly.js/commit/9aa5494f43869a7f1355267458c91b22fba8fe59.

emilykl commented on src/components/shapes/draw.js:
```diff
@@ -605,13 +607,30 @@ function drawLabel(gd, index, options, shapeGroup) {
+    var text = options.label.text;
+    if(options.label.texttemplate) {
+        var templateValues = {};
+        if(options.type !== 'path') {
```
Makes sense.

I think we still need the check here though, because a texttemplate might get applied to a new path drawn by the user if `newshape.texttemplate` is set.

Or can we prevent the texttemplate from being applied to the new shape if the user draws a path?

archmoj commented:
@emilykl the PR is looking good now.
In your tests, would you possibly add/modify a subplot to use category axes as well?
Also it would be great to have a paper referenced shape in the mock.

alexcjohnson commented on src/components/shapes/draw.js:
```diff
@@ -605,13 +607,30 @@ function drawLabel(gd, index, options, shapeGroup) {
+    // If no label text or texttemplate, return
+    if(!(options.label.text || options.label.texttemplate)) return;
+
+    // Text template overrides text
```
Let `defaults` do the overriding - ie don't coerce `text` if there's a `texttemplate`.

alexcjohnson replied:
and even better - don't try to coerce `texttemplate` if `type==='path'` - then you don't need the `if(options.type !== 'path')` below :)

emilykl replied:
@alexcjohnson Changed the coercion in `defaults.js` but I believe we still need the `if(texttemplate)` statement (see updated code) -- correct me if wrong!

alexcjohnson commented on src/components/shapes/draw.js:
```diff
@@ -605,13 +607,30 @@ function drawLabel(gd, index, options, shapeGroup) {
+        if(options.type !== 'path') {
+            var _xa = Axes.getFromId(gd, options.xref);
+            var _ya = Axes.getFromId(gd, options.yref);
+            Object.keys(shapeLabelTexttemplateVars).forEach(function(key) {
+                var val = shapeLabelTexttemplateVars[key](options, _xa, _ya);
+                if(val !== undefined) templateValues[key] = val;
+            });
```
```suggestion
            for (var key in shapeLabelTexttemplateVars) {
                var val = shapeLabelTexttemplateVars[key](options, _xa, _ya);
                if(val !== undefined) templateValues[key] = val;
            }
```
A little simpler and faster 😎

alexcjohnson commented on src/components/shapes/label_texttemplate.js:
```diff
@@ -0,0 +1,82 @@
+
+function slopeFn(shape, xa, ya) {
+    return (shape.type !== 'line') ? undefined : (
+        (d2l(shape.y1, ya) - d2l(shape.y0, ya)) /
+        (d2l(shape.x1, xa) - d2l(shape.x0, xa))
```
```suggestion
        dyFn(shape, xa, ya) / dxFn(shape, xa)
```

archmoj commented on test/image/mocks/zz-text_on_shapes_texttemplate.json:
```diff
@@ -0,0 +1,357 @@
+            "eraseshape"
+        ]
+    }
+}
```
There are some unused keys in this mock which cause the `test-mock` to fail.
Please see https://app.circleci.com/pipelines/github/plotly/plotly.js/8509/workflows/d2ea2360-d4de-410b-bb19-40ba0a3d9fc4/jobs/188326.

Also you could simply test it using the following command:
```sh
npm run test-mock zz-text_on_shapes_texttemplate
```

archmoj commented on test/image/mocks/zz-text_on_shapes_texttemplate.json:
```diff
@@ -212,16 +211,6 @@
-                "editable": true,
-                "path": "M3.71,0.75L2.26,3.99L3.71,3.99Z",
-                "layer": "above"
-            },
```
OK. Please update the baseline so that image test pass: https://app.circleci.com/pipelines/github/plotly/plotly.js/8520/workflows/491afb80-1892-4f25-8f62-b5b25db1bfb4/jobs/188422/artifacts

emilykl commented:
@alexcjohnson @archmoj I think this PR is finally ready! I've addressed all above conversations -- let me know if there's anything additional remaining.

I added a section in the PR description to explain the behavior for `log`, `date`, and `category` axes, and I also added a category axis and a paper-referenced shape to the shape texttemplates mock.

archmoj commented on src/lib/index.js:
```diff
@@ -1145,6 +1176,14 @@ function templateFormatString(string, labels, d3locale) {
+                '*': (function(v) { return v * parsedNumber; }),
+                '/': (function(v) { return v / parsedNumber; }),
+            }[parsedOp](value);
+        }
```
Wondering if we could simplify this logic without those functions? Something like: 
```js
        if(value !== undefined)  {
            if(parsedOp === '*') value *= parsedNumber;
            if(parsedOp === '/') value /= parsedNumber;
        }
```

emilykl replied:
Done!

archmoj commented:
All is looking :star: good :star: to me.
@alexcjohnson would you like to have a final look?

alexcjohnson reviewed (approved):
💃 Fantastic work @emilykl!

mosaikme commented:
Hey thanks for this Future, i try to have something that can calculate percentage distance,  so simply draw a rectangle, than ( (y0 / y1) -1  ) *100  , can we intigrate something like this,  or  able do make math direct in the texttemplate.  Or somebody know of a other way for it?

alexcjohnson commented:
@mosaikme arbitrary math was not implemented but was discussed in the original issue https://github.com/plotly/plotly.js/issues/6511#issuecomment-1462937672 - feel free to open a new issue to request this, or even better a PR to make it happen 😉

mosaikme commented:
ok , sad. I will try to add this. But where to start, Would it be enough to edit or add percent calculation here https://github.com/plotly/plotly.js/blob/master/src/components/shapes/label_texttemplate.js as a new function , and than add it the module.export?
